google/cloud/storage/async_reader.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/async_reader.h"
+#include "google/cloud/storage/async_reader_connection.h"
+#include "google/cloud/internal/make_status.h"
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+StatusOr<std::pair<ReadPayload, AsyncToken>> InvalidTokenError(
+    internal::ErrorInfoBuilder eib) {
+  return internal::InvalidArgumentError("invalid token", std::move(eib));
+}
+
+StatusOr<std::pair<ReadPayload, AsyncToken>> ClosedStreamError(
+    internal::ErrorInfoBuilder eib) {
+  return internal::CancelledError("closed stream", std::move(eib));
+}
+
+class Discard : public std::enable_shared_from_this<Discard> {
+ public:
+  explicit Discard(std::unique_ptr<AsyncReaderConnection> impl)
+      : impl_(std::move(impl)) {}
+
+  void Loop() {
+    auto self = shared_from_this();
+    self->impl_->Read().then([self](auto f) mutable {
+      auto s = std::move(self);
+      auto response = f.get();
+      if (absl::holds_alternative<Status>(response)) {
+        s.reset();
```

Comment:
Wouldn't this happen during the `return` anyway?

Reply:
Done.